### PR TITLE
OPT: Bypass `submodule add` when adding subdatasets

### PR DESCRIPTION
**Pull request by mih**
Instead go manual, avoiding various checks, because we already
know inside `save` that a particular directory is considered untracked
by Git.

This approach is substantially faster. Already with 50 subdatasets
the speed-up is 2x.

Quick'n'dirty benchmark:

```
% rm -rf * .git* .datalad; datalad create . && \
  ( for i in $(seq 50); do datalad create $i; done ) && \
  time datalad save && datalad status
```

Current master

```
datalad save  3.70s user 1.50s system 103% cpu 4.998 total
```

This PR

```
datalad save  1.68s user 1.02s system 108% cpu 2.499 total
```

It may well be that it needs more safety checks that would eat up that
margin. But in the context of
https://github.com/datalad/datalad/issues/2880#issuecomment-667041918
(save thousands of new subdatasets) even much smaller benefits would
still translate to substantial absolute speed-ups.

Thx again to @kyleam for digging out that `update-index` trick.

TODO:

- [x] consolidate config and `.gitmodules` manipulation into a single action for all added submodules at once
- [x] RF helpers into documented standalone functions with tests

### CODE REVIEW

**Comment by mih** [comment]
This works now. The only test failure is:

```
======================================================================
1649ERROR: datalad.distribution.tests.test_install.test_install_recursive
1650----------------------------------------------------------------------
1651Traceback (most recent call last):
1652  File "/home/travis/virtualenv/python3.5.6/lib/python3.5/site-packages/nose/case.py", line 134, in run
1653    self.runTest(result)
1654  File "/home/travis/virtualenv/python3.5.6/lib/python3.5/site-packages/nose/case.py", line 152, in runTest
1655    test(result)
1656AssertionError: Test was too slow (took 31.0307s, threshold was 30.0000s)
```

it is unclear to me how it can get slower (I have not yet observed a scenario where things got slower), but there are some optimization that can still be done.

**Comment by mih** [comment]
Update: Consolidation on applying all `.gitmodules` and `.git/config` changes at once speeds things up by another 40%, leading to a 3x overall speed-up for 50 subdatasets.

File: datalad/support/gitrepo.py [hunk 1]
```diff
@@ -4076,6 +4056,102 @@ def _save_add(self, files, git_opts=None):
...
+                    'update-index', '--add', '--replace', '--cacheinfo', '160000',
+                    i['commit'], i['rpath']
+                ])
+                gmprops = dict(path=i['rpath'], url=i['url'])
+                if i['id']:
+                    gmprops['datalad-id'] = i['id']
+                write_submodule_config(
+                    gmf, i['rpath'], gmprops)
+                write_submodule_config(
+                    gcf, i['rpath'], dict(active='true', url=i['url']))
```
Comment (kyleam) [inline datalad/support/gitrepo.py:4140]: With this switch to writing the `.git/config` entries on our own, there's a change in how the `url` is represented: before it was an absolute path and now it is relative.  I think that should be fine and for our purposes probably preferable.

File: datalad/support/gitrepo.py [hunk 2]
```diff
@@ -4076,6 +4056,102 @@ def _save_add(self, files, git_opts=None):
...
+
+                # This mirrors the result structure yielded for
+                # to_stage_submodules below.
+                yield get_status_dict(
+                    action='add',
+                    refds=self.pathobj,
+                    # MIH: really? why not 'dataset', it seems "add" of an existing
+                    # subdataset (i.e. update) does that, and this one mimiks it
+                    # still, two wrongs don't make a right.
```
Comment (kyleam) [inline datalad/support/gitrepo.py:4149]: How to handle this was brought up explicitly at the time.

https://github.com/datalad/datalad/pull/3398#discussion_r282928154

Ideally I think we'd switch to using `type=dataset` for adding both modified and new datasets.  The main obstacle I see is that the result values are exposed via result hooks, so we need to worry about breakage.

In my view this comment doesn't add any value, though.

Comment (mih) [reply datalad/support/gitrepo.py:4149]: I will replace this comment with a link to this statement

File: datalad/support/gitrepo.py [hunk 3]
```diff
@@ -4076,6 +4056,102 @@ def _save_add(self, files, git_opts=None):
...
+            v = v.replace('\\', '\\\\')
+            if section:
+                # must not have additional unquoted quotes
+                v = v.replace('"', '\\"')
+            elif v and (v[0] in white or v[-1] in white):
+                # quoting the value due to leading/trailing whitespace
+                # prevent inner unquotes quotes
+                v = v.replace('"', '\\"')
+                v = '"{}"'.format(v)
+            return v
```
[the remark this answers is not in the document]
Comment (mih) [reply datalad/support/gitrepo.py:4086]: Thx! Will change accordingly.

[the remark this answers is not in the document]
Comment (mih) [reply datalad/support/gitrepo.py:4086]: This is done now.

File: datalad/support/gitrepo.py (hunk 1 above)
Comment (mih) [reply datalad/support/gitrepo.py:4140]: True. I had the same feeling of "should be better". Related: https://github.com/datalad/datalad/pull/4025#issue-362968783

**Comment by mih** [comment]
Pushed update, rebased on current master

File: datalad/support/gitrepo.py (hunk 1 above)
Comment (mih) [reply datalad/support/gitrepo.py]: In any case the issue from #3538 remains here, as something earlier makes the path absolute:

```
% dl create raw
% dl create bids
% cd bids
bids % datalad install -d . -s ../raw sourcedata --reckless
bids % cat .gitmodules
[submodule "sourcedata"]
        path = sourcedata
        url = /tmp/d1/bids/../raw
        datalad-id = ca96f420-2f90-4f9c-8c46-0e6213dcfd99
```

**Comment by mih** [comment]
Related: #4853

**Review by bpoldrack — COMMENTED** [review]
LGTM

**Comment by mih** [comment]
Thx all!

**Pull request merged**